### DISCUSSION TITLE
tutorials/navigation/navigation_using_navigationagents

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#12] [User-contributed notes]
# tutorials/navigation/navigation_using_navigationagents

NavigationsAgents are helper nodes that combine functionality for pathfinding, path following and agent avoidance for a Node2D/3D inheriting parent node. They facilitate common calls to the Navigat...

https://docs.godotengine.org/en/latest/tutorials/navigation/navigation_using_navigationagents.html

**Comment by RetrocadeMedia** [comment]
map_get_iteration_id not found in base GDScriptNativeClass

**Reply by Calinou** [reply]
Which Godot version are you using? The method exists as of 4.3 (and also in 4.2):

https://github.com/godotengine/godot/blob/d40fc50f086d14583c7cc979ed4e5363ac223717/doc/classes/NavigationServer2D.xml#L507-L514

https://github.com/godotengine/godot/blob/d40fc50f086d14583c7cc979ed4e5363ac223717/doc/classes/NavigationServer3D.xml#L571-L578

**Comment by boholstapp** [comment]
video Example
https://youtu.be/qIqcp7xBGkw

**Comment by Keyvan-Rostyslav** [comment]
Ok, how to add Gravity?

**Reply by elementbound** [reply]
> The movement is entirely in the hands of users and their custom scripts.

You can combine the navigation agent's output with your own movement code, e.g. with the `CharacterBody3D` script template.

**Reply by cmaguranis** [reply]
In addition to what @elementbound has said, in 3D a problem I ran into and will need to be accounted for is ensuring that the Y value of the velocity is not lost when `_on_velocity_computed` is called:

For example, in CharacterBody3D:
```gdscript
func _on_velocity_computed(safe_velocity: Vector3):
    safe_velocity.y = velocity
	velocity = safe_velocity
	move_and_slide()
```

**Comment by stephen-polyxia** [comment]
I think the NavigationAgent template may be incorrect, it includes delta twice in the movement computation. Once when setting newVelocity, and then a second time in OnVelocityComputed.

In theory, Velocity should not include time in the computation at all. I assume this will be handled by the NavigationServer.

Corrected template elements:
In _PhysicsProcess:
Vector2 newVelocity = GlobalPosition.DirectionTo(nextPathPosition)

In OnVelocityComputed
GlobalPosition = GlobalPosition.MoveToward(GlobalPosition + safeVelocity)

Should be sufficient.

**Comment by linky00** [comment]
Really important to note that when calculating whether the path has been reached, the `path_desired_distance` is relative to the actual global position of the agent, _not_ their 'virtual' position on the navmesh. This means that if the navmesh is floating any higher off the ground than the `path_desired_distance`, the agent will never be able to reach that point!

Moreover, it seems **the `path_height_offset` affects the calculation of distance from the path**. This means you can compensate for the offset using `path_height_offset`.

I was banging my head against this for ages. Watch out for a short `path_desired_distance`!
